### PR TITLE
Integrate second pass ex 21 model improvements and move to notebook based training

### LINKED ISSUE #78: Second Pass Ex. 21 Model Improvements
### Overview

Spinning off the unresolved tasks from #68 into a separate issue. Also, post-Dagsterization, some of these changes become easier to implement in a separate PR from the original, so it made sense to break up into a second pass issue.

Example of filing that should fail and not create a PDF:
* 92416-0000892569-94-000102
Example of filing that isn't rendered well with HTML:
* 9342-0000009342-95-000008

Example of filing with "footnotes":
103872-0001193125-13-444053

```[tasklist]
### Next steps
- [ ] Use Corpwatch dataset for further validation
- [x] More comparison of validation and extracted dataframes with more flexible ergonomic improvements to `GCSArchive`
- [x] Add an x and y threshold for distinct subsidiary cutoffs
- [x] Don't create PDFs for empty tables or Ex. 21 tables that can't be rendered, skip instead. Also handle "crossed out" filings
- [x] Zach: Look into what's happening with HTMLs that are rendered without table structure or in "paragraph" form
- [x] Look at the layouts that are most commonly "missed"
- [x] Try taking out "paragraph" layout docs and having same filer from different years replace that entry
- [ ] Nice to have: breakout `layoutlm-finetune` into ops
- [ ] Nice to have: Try clustering the final hidden states instead of using heuristic based table extractor
- [ ] Nice to have: Exclude anything below "Footnotes" or similar keywords
```

### PR DESCRIPTION
**Pull request by zschira**
# Overview

Closes #78.

This PR adds improvements to the exhibit 21 extractor and moves the model implementation and training to a notebook which can be managed by dagster.

#### What did you change in this PR?

- Move exhibit 21 model to a notebook
- Pull model improvements from @katie-lamb's branch [second-pass-ex21-improvements](https://github.com/catalyst-cooperative/mozilla-sec-eia/tree/second-pass-ex21-improvements)
- Add Exhibit 21 layout classifier model
- Integrate layout classifier to production pipeline to enable filtering paragraph docs from record linkage

### CODE REVIEW

File: src/mozilla_sec_eia/library/generic_io_managers.py [hunk 1]
```diff
@@ -19,3 +21,19 @@ def load_from_path(self, context: InputContext, path: UPath) -> pd.DataFrame:
         """Read parquet."""
         with path.open("rb") as file:
             return pd.read_parquet(file)
+
+
+class PickleUPathIOManager(UPathIOManager):
```
Comment (zschira) [inline src/mozilla_sec_eia/library/generic_io_managers.py:26]: I added this to save pickled asset outputs to GCS. This is needed because I separated the ex21 inference dataset creation from actually running the model, but the datasets take up too much space if they're saved locally.

File: src/mozilla_sec_eia/library/mlflow/__init__.py [hunk 2]
```diff
@@ -12,6 +16,22 @@
 )
 
 
+def pyfunc_model_asset_factory(name: str, mlflow_run_uri: str):
```
Comment (zschira) [inline src/mozilla_sec_eia/library/mlflow/__init__.py:19]: This function will create an asset to load a model from mlflow. Using `create_model` is a little bit of a weird way to provide configuration to the asset, but this ensures that the default value for `mlflow_run_uri` will show up in the dagster UI.

File: src/mozilla_sec_eia/models/sec10k/__init__.py [hunk 3]
```diff
@@ -37,39 +48,76 @@
...
-    concurrency_limit=4,
+    tag_concurrency_limits=[
+        {"key": "model", "value": "exhibit21_extractor", "limit": 2},
+    ],
 )
 
-ex21_validation_job = model_jobs.create_validation_model_job(
-    "ex21_extraction_validation",
-    ex_21.validation_assets,
+
```
Comment (zschira) [inline src/mozilla_sec_eia/models/sec10k/__init__.py:56]: At some point I'd like to clean up all this asset/job creation stuff, but I don't think this is the best time for that.

File: src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_extractor.ipynb [hunk 4]
```diff
@@ -0,0 +1,1144 @@
+{
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_extractor.ipynb:1]: <p>At a high level, I think I'm confused about whether this notebook should be used to run only validation, or also inference on the whole dataset. If it's only used for validation, then I think the job name and/or the name of the notebook should be changed to include "validation" in the name. It seems like this notebook is used for the&nbsp;<code>ex21_training</code>&nbsp;job and not&nbsp;<code>ex21_extraction</code>&nbsp;, but it's called&nbsp;<code>ex21_extraction.ipynb</code></p><p> </p><p>Also, maybe should be "This notebook implements a model built on top of&nbsp;<a href="https://huggingface.co/microsoft/layoutlmv3-base/tree/main" target="_blank">layoutlmv3</a>&nbsp;<strong>to extract tables</strong>&nbsp;from Exhibit 21 attachments to SEC-10k filings." I think the current wording doesn't make sense to me.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_extractor.ipynb:1]: <p>I think these upstream assets are hard to differentiate. Took a stab at making it a little more verbose:</p><ul><li><code>ex21_training_data</code>: Dataset of labeled Ex. 21 documents produced in Label Studio and used to train&nbsp;<code>layoutlm</code>&nbsp;. Each word in a document has an IOB format entity tag indicating one of these classes: subsidiary, location of incorporation, ownership percentage, or other.</li><li><code>ex21_validation_set</code>: Transcribed tables from Ex. 21 documents describing the expected inference output on a validation set of filings.</li><li><code>ex21_failed_parsing_metadata</code>: Metadata for any validation filings that couldn't be parsed and included in the inference dataset (usually empty)</li><li><code>ex21_inference_dataset</code>&nbsp;: Parsed filings prepped for the inference model. If running validation, this should be the validation set of filings.</li></ul><p> </p><p>If running inference on all the docs, do you still need the validation set in&nbsp;<code>ex21_validation_set</code>&nbsp;to be materialized? If you are running validation, should&nbsp;<code>ex21_inference_dataset</code>&nbsp;just contain the validation set of filings?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_extractor.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.        "layoutlm_training_run": "layoutlm-labeledv0.2",
</pre><p>Does setting a different value in the Dagster config overrule what's set in this cell?</p><p> </p><p>What actually does this string represent? Is it the model path from MLflow, something from Dagster, or something local? I see&nbsp;<code>exhibit21_extractor</code>&nbsp;and&nbsp;<code>layoutlm_extractor</code>&nbsp;on the MLFlow registered models, but neither seems right.</p><p> </p><p>Should this be&nbsp;<code>layoutlm_training_run</code>&nbsp;or&nbsp;<code>layoutlm_uri</code>? The documentation cell above says&nbsp;<code>layoutlm_uri</code>&nbsp;but maybe just needs to be updated?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_extractor.ipynb:1]: <p>Maybe would add this: "Model finetuning will only be run if configured to do so (<code>layoutlm_training_run = None</code>&nbsp;) , otherwise a pretrained version will be used from the&nbsp;<code>mlflow</code>&nbsp;tracking server.</p><p>Also I think in step 2 we should say Named Entity Recognition (NER) since it's the first place we use NER.</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_extractor.ipynb:1]: <p>Maybe "labeled validation data" should be "manually transcribed validation tables"? Just to differentiate between this and the Label Studio training data?</p><p> </p><p>Also, if this inference section is used in the ex21_extraction job, then maybe mention that is what's run to actually extract tables, and not just for validation.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_layout_classifier.ipynb [hunk 5]
```diff
@@ -0,0 +1,337 @@
+{
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_layout_classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #11.        def predict(self, context, model_input: pd.DataFrame):
</pre><p>I assume we use predict when we pull down the classifier during an inference run and actually classify documents?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/notebooks/exhibit21_layout_classifier.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #25.    for classifier, model in classifiers.items():
</pre><p>Do you think we should run cross validation of the models every time? Or should we just choose one (SVM performed the best). I guess it's fine that we just log all of them in MLFlow and then just choose one but could potentially use the wrong one in production.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/catalyst-cooperative/mozilla-sec-eia/pull/95/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

**Comment by katie-lamb** [comment]
I changed the feature creation of the paragraph layout classifier slightly and left some comments on the Ex. 21 extraction notebook.

I think what I'm most unclear about in the near structure is what the difference is between the `ex21_training` and `ex21_extraction` job is. The `ex21_training` job has the option of fine-tuning LayoutLM and then running inference on the validation set, whereas the `ex21_extraction` job just loads a model from a run and performs inference on the whole dataset? I felt confused where to set the config in Dagster and where to set the config in the notebook. Also, I wasn't sure where model paths were coming from, since they no longer seem to be coming from the MLFlow model registry.

Maybe we need to include a summary of the different jobs in the README?

**Review by katie-lamb — CHANGES_REQUESTED** [review]
I think it might be worth getting on the phone to talk about how the Dagster config stuff is now set up. Nothing seems broken for me or like it needs major changes. I integrated a few more changes to the model pipeline from my second pass improvements branch, but otherwise I think all the changes from there are integrated in this PR.

**Comment by katie-lamb** [comment]
Oh also, I'm not able to execute the op `ex21_training_data`, I'm getting a KeyError on "labels" in `format_label_studio_output`

File: src/mozilla_sec_eia/models/sec10k/ex_21/data/common.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    return cik in tracking_df.CIK.unique()
+
+
+def format_label_studio_output(
+    labeled_json_dir: Path,
+    pdfs_dir: Path,
+) -> pd.DataFrame:
+    """Format Label Studio output JSONs into dataframe."""
+    labeled_df = pd.DataFrame()
+    tracking_df = validation_helpers.load_training_data("ex21_labels.csv")
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/mozilla_sec_eia/models/sec10k/ex_21/data/common.py]: I renamed this file to `ex21_labeled_filings.csv` but feel free to change. Also, I pulled over the version from my other branch which has paragraph layout filings commented out.

File: src/mozilla_sec_eia/models/sec10k/ex_21/ex21_validation_helpers.py [hunk 7]
```diff
@@ -1,13 +1,132 @@
...
+            "float64", errors="ignore"
+        )
+    # drop rows that have a null subsidiary value
+    extracted_df = extracted_df.dropna(subset="subsidiary")
+    return extracted_df
+
+
 def clean_ex21_validation_set(validation_df: pd.DataFrame):
     """Clean Ex. 21 validation data to match extracted format."""
+    validation_df = remove_paragraph_tables_from_validation_data(validation_df)
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/ex_21/ex21_validation_helpers.py:129]: I added this, to automatically remove paragraph layout docs from the validation data (even though the current version of the validation data already has them removed). I figured it couldn't hurt, but you can take it out if you want.
